test/test_model.py

```diff
@@ -3391,17 +3395,44 @@ def test_opened_ports_warnings(self):
         self.assertIsInstance(ports_set, set)
         ports = sorted(ports_set, key=lambda p: (p.protocol, p.port))
         self.assertEqual(len(ports), 2)
-        self.assertIsInstance(ports[0], ops.OpenedPort)
+        self.assertIsInstance(ports[0], ops.Port)
         self.assertEqual(ports[0].protocol, 'tcp')
         self.assertEqual(ports[0].port, 8080)
-        self.assertIsInstance(ports[1], ops.OpenedPort)
+        self.assertIsInstance(ports[1], ops.Port)
         self.assertEqual(ports[1].protocol, 'udp')
         self.assertEqual(ports[1].port, 1000)
 
         self.assertEqual(fake_script_calls(self, clear=True), [
             ['opened-ports', ''],
         ])
 
+    def test_set_ports(self):
+        fake_script(self, 'open-port', 'exit 0')
+        fake_script(self, 'close-port', 'exit 0')
+        fake_script(self, 'opened-ports', 'exit 0')
+        self.unit.set_ports(8000, 8025)
+        calls = fake_script_calls(self, clear=True)
+        self.assertEqual(calls.pop(0), ['opened-ports', ''])
+        # We make no guarantee on the order the ports are opened.
+        calls.sort()
+        self.assertEqual(calls, [
+            ['open-port', '8000/tcp'],
+            ['open-port', '8025/tcp'],
+        ])
+        fake_script(self, 'opened-ports', 'echo 8025/tcp')
```

Comment:
just for test consistency, it seems like `opened-ports` should return both ports, not just one of them. Otherwise, the 'side effect' of the first calls isn't actually represented.

Reply:
The test is setting a fresh state (via `fake_script`) each time rather than trying to tell a single story from start to finish (it did start out that way, but I found I was doing extra work just to have a single story, so stripped that out). If one story all the way through the test is preferred, I can adjust to that.